sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveQl.scala

```diff
@@ -1051,6 +1060,168 @@ https://cwiki.apache.org/confluence/display/Hive/Enhanced+Aggregation%2C+Cube%2C
       throw new NotImplementedError(s"No parse rules for:\n ${dumpTree(a).toString} ")
   }
 
+  protected val windowDefs = new ThreadLocal[Map[String, Seq[ASTNode]]] {
+    override def initialValue() = Map.empty[String, Seq[ASTNode]]
+  }
```

Comment:
It would be great if we can remove this `ThreadLocal`. But if it requires too much work, I guess it's OK to leave it as is. `ThreadLocal` can always be error prone.